### DISCUSSION TITLE
Appium unable connect to Remote Android Emulator (docker container on a server)

### DISCUSSION THREAD

**Discussion by NikolaRusano** [appium/appium#17088] [Q&A]
Hello everybody! Can Anybody help me with to connect maven Java autotesting project to connect remote started docker Android Emulator Samsung Galaxy S6 ? Te emulator works on remote server as a selenoid. But I can’t concect to it - everytime I tried - i got error message like:
Unable to create a new remote session. Please check the server log for more details. Original error: The session identified by 0518e9c8-6b49-4645-894e-9f41e1600ce9 is not known
or
Could not start a new session. Possible causes are invalid address of the remote server or browser start-up failure.

I used RemoteWebDriver for connect and Android Driver. In both ocassions it doesn’t give any results - the tests which correctly works on local Android emulators - on remote emulator driver even not creates and function after AppiumDriver flows processes (Http commands and operations) retuns null instead AdnroidDriver and RemoteWebDriver.
I use the capabilities (here the RemoteWebDriver used but I used the Android Driver too):

DesiredCapabilities dcap = new DesiredCapabilities();

    dcap.setCapability(MobileCapabilityType.DEVICE_NAME, "samsung_galaxy_s6_8.1");
    dcap.setCapability(MobileCapabilityType.UDID, "emulator-5554");
    dcap.setCapability(MobileCapabilityType.PLATFORM_NAME, "Android"); //platformName
    dcap.setCapability(MobileCapabilityType.PLATFORM_VERSION, "8.1.0"); //platformVersion
    dcap.setCapability(MobileCapabilityType.AUTOMATION_NAME, "UiAutomator2");
    dcap.setCapability("enableVNC", true);
    dcap.setCapability("enableVideo", false);

    dcap.setCapability("app", "http://guest:guest@autobuild.fozzy.lan/repository/download/Mobile_MAppFora_Android_ForaTestApkBuild/273216:id/foraapp/build/outputs/apk/googleMtest/fora/ForaMobileAndroid-1.23.2-google-mtest-fora.apk");
    dcap.setCapability("appPackage", "ua.fora.android.mtest");
    dcap.setCapability("appActivity", "ua.fora.android.ui.activity.SplashActivity");
    dcap.setCapability("unicodeKeyboard", true);
    dcap.setCapability("resetKeyboard", true);
    dcap.setCapability("skipUnlock", true);
    dcap.setCapability("ensureWebviewsHavePages", true);
    dcap.setCapability("newCommandTimeout", 120);
    dcap.setCapability("nativeWebScreenshot", true);
    dcap.setCapability("acceptInsecureCerts", true);
    dcap.setCapability("browserstack.gpsLocation", "50.45466,-30.5238");
    dcap.setCapability("gpsEnabled", "true");
    dcap.setCapability("language", "ukr");
    dcap.setCapability("locale", "UA");

    try {
            driver =  new RemoteWebDriver(new URL("http://sftv-auto001.fz.fozzy.lan:49238/wd/hub"), dcap); // addres Remote Android Emulator

        pause(8000);
        ((RemoteWebDriver) driver).setFileDetector(new LocalFileDetector());
        System.out.println("Start Driver");
    } catch (Exception e) {
        e.printStackTrace();
        e.getMessage();
    }
    return driver;
And as a result of this code I am getting null instead of driver and it’s settings. Could anybody help to connect ?

![image](https://user-images.githubusercontent.com/36786249/175271452-0314db1f-1958-47b8-ab81-40027c9cec68.png)

The same capabilities gives opportunity correctly connect with Python but in Java it gives errors

**Comment by arthuRHD** [comment]
You need to expose the ADB port of your container, for example : `-p 5037:5037` if you're using the default port.

It's because your local appium server is searching for an ADB server inside your hosted server in order to create a socket. But your emulator which comes with Android SDK tools is sandboxed inside a container, therefore you need to expose your container port in your hosted server to access the ADB server inside the container.
